### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by nickhsmith**
## PR checklist

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
    - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
    - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/sarek/tree/master/.github/CONTRIBUTING.md)
    - [ ] If necessary, also make a PR on the nf-core/sarek _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint .`).
- [X] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

This PR seeks to fix 2 problems.
1) For each module defined in conf/modules.config set an appropriate publishDir list containing the `path`, `mode`, `pattern`, and `enabled` keywords. This should now address #471 and match the module stle
2) For the many of the variant_calling steps, there were some typos and function argument ordering errors. These have been fixed.

### CODE REVIEW

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:134]: if it is not enabled, i don't think this needs to be here

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:153]: same here

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:164]: same

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:198]: see above, and the `path` could actually also be removed

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:207]: we are inconsistent here when we publish. Shouldn't be necessary

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:264]: see above

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:308]: This was commented out, because @GCJMackenzie is adding in a new subworkflow that also allows joint calling. @GCJMackenzie maybe you can take a look at the gatk based tools and their parameters.

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:399]: this is part of `ext.container` now
```suggestion
```

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:411]:
```suggestion
```

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:423]: any reason why you would not want to publish here?

File: .gitignore
Comment (FriederikeHanssen) [inline .gitignore:10]: is there a reason you added this?

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:14]:
```suggestion
// Generic process options for all workflows
```

I think this comes from the template, so maybe we can leave it in like this

File: conf/modules.config
Comment (FriederikeHanssen) [inline conf/modules.config:17]: Maybe lets keep the default publishing strategy, or did you experience difficulties with it?

File: docs/output.md
Comment (FriederikeHanssen) [inline docs/output.md:145]:
```suggestion
    - `TSV` files to start `Sarek` from `preparerecalibration`, `recalibrate` or `variantcalling` steps.
```

Steps are renamed in the new version

File: docs/output.md
Comment (FriederikeHanssen) [inline docs/output.md:147]:
```suggestion
    - `TSV` files to start `Sarek` from `preparerecalibration`, `recalibrate` or `variantcalling` steps for a specific sample.
```

File: docs/output.md
Comment (FriederikeHanssen) [inline docs/output.md:158]:
```suggestion
    - `TSV` files to start `Sarek` from `preparerecalibration`, `recalibrate` or `variantcalling` steps.
```

File: docs/output.md
Comment (FriederikeHanssen) [inline docs/output.md:171]:
```suggestion
    - `TSV` files to start `Sarek` from `variantcalling` step.
```

File: docs/output.md
Comment (FriederikeHanssen) [inline docs/output.md:173]:
```suggestion
    - `TSV` files to start `Sarek` from `variantcalling` step for a specific sample.
```

File: nextflow.config
Comment (FriederikeHanssen) [inline nextflow.config:29]: there is an open PR currently discussing using null or false, maybe we can wait ehre until that one is merged.

File: nextflow_schema.json
Comment (FriederikeHanssen) [inline nextflow_schema.json:667]: in that case this needs to stay in

File: modules/local/gatk4/haplotypecaller/main.nf
Comment (FriederikeHanssen) [inline modules/local/gatk4/haplotypecaller/main.nf:11]: the gatk4 modules are in the process of being pulled from nf-core/modules. we should reflect any changes from there, we need double check with the upstream modules

File: workflows/sarek.nf
Comment (FriederikeHanssen) [inline workflows/sarek.nf:321]:
```suggestion
    if (params.step in ['mapping', 'preparerecalibration', 'recalibrate']) {
```

File: workflows/sarek.nf
Comment (FriederikeHanssen) [inline workflows/sarek.nf:276]:
```suggestion
    if (params.step in ['mapping', 'preparerecalibration']) {
```

File: workflows/sarek.nf
Comment (FriederikeHanssen) [inline workflows/sarek.nf:271]:
```suggestion
    if (params.step == 'preparerecalibration') {
```

File: workflows/sarek.nf
Comment (FriederikeHanssen) [inline workflows/sarek.nf:53]:
```suggestion
        case 'variantcalling':       csv_file = file("${params.outdir}/preprocessing/csv/recalibrated.csv",            checkIfExists: true); break
```

File: workflows/sarek.nf
Comment (FriederikeHanssen) [inline workflows/sarek.nf:51]:
```suggestion
        case 'preparerecalibration': csv_file = file("${params.outdir}/preprocessing/csv/markduplicates_no_table.csv", checkIfExists: true); break
```

File: subworkflows/local/prepare_genome.nf
Comment (FriederikeHanssen) [inline subworkflows/local/prepare_genome.nf:86]:
```suggestion
    if (!(params.known_indels_tbi) && params.known_indels && ('mapping' in step || 'preparerecalibration' in step)) {
```

File: subworkflows/local/prepare_genome.nf
Comment (FriederikeHanssen) [inline subworkflows/local/prepare_genome.nf:72]:
```suggestion
    if (!(params.dbsnp_tbi) && params.dbsnp && ('mapping' in step || 'preparerecalibration' in step || 'controlfreec' in tools || 'haplotypecaller' in tools|| 'mutect2' in tools || 'tnscope' in tools)) {
```

**Review by FriederikeHanssen — CHANGES_REQUESTED** [review]
hey @nickhsmith thanks a lot for fixing the publish copy issue 😊 anything that is not right in `nf-core/modules` should be fixed in that repository, so we can properly pull from there. Since you have also adressed some of the null/false things in the nextflow.config here, maybe we can wait for  conclusion on the discussion and Maxime to merge his PR. We could certainly add the things about Haplotypecaller here, and once Gavin is ready just overwrite it.

**Comment by FriederikeHanssen** [comment]
Before we can merge for sure, need to get the CI tests back to pass as well as the linting for sure 🙈

File: tests/config/nextflow.config
Comment (FriederikeHanssen) [inline tests/config/nextflow.config:3]: I think this might need to stay in for the CI tests to pass

File: .gitignore
Comment (nickhsmith) [reply .gitignore:10]: I guess that was for my own sanity. Using vim as an editor creates these temporary files. I can remove it

File: docs/output.md
Comment (nickhsmith) [reply docs/output.md:145]: new version of what? The sarek.nf workflow uses variantcalling, but no longer includes the '_' handling/concatenation.

File: nextflow.config
Comment (nickhsmith) [reply nextflow.config:90]: do you mean having the `conf/modules.config` file draw from a `"${params.publish_dir_mode}" `or something?

File: conf/modules.config
Comment (nickhsmith) [reply conf/modules.config:134]: I like having the same format, and if someone decides tomorrow to output/publish this module. All they need to do is to change `enabled` But it's definitely not used now

**Comment by maxulysse** [comment]
You can check https://github.com/nf-core/sarek/pull/468 for some updates, but we're probably going to get another PR whenever the tools release is happening (which I guess should be soon now)

File: lib/WorkflowMain.groovy
Comment (maxulysse) [inline lib/WorkflowMain.groovy:79]: I would not remove that

File: workflows/sarek.nf
Comment (maxulysse) [inline workflows/sarek.nf:383]: looks like merge confilcts

**Comment by maxulysse** [comment]
Still some comments not addressed, but looking better and better, I like the update on concat_vcf

File: lib/WorkflowMain.groovy
Comment (nickhsmith) [reply lib/WorkflowMain.groovy:79]: we need to have some work around for utilizing the automatic restart. This check totally negates the checks at `workflows/sarek` lines 46 - 56

Comment (FriederikeHanssen) [reply lib/WorkflowMain.groovy:79]: hm ok I would say we need to fix sarek not the template then, what do you think?

**Comment by nickhsmith** [comment]
This whole PR is still not right or up to date. But There is good conversation here.

File: lib/WorkflowMain.groovy
Comment (nickhsmith) [reply lib/WorkflowMain.groovy:79]: I think it's quite a useful feature to be able to restart the analysis. There is already a warning, and if the step isn't there it throws an error. You need to designate a step, which essentially is a shortcut for editing the config file. just set step = `variant_calling` instead of `input = 'Path/To/Dir/preprocessing/csv/recalibrated.csv'`

Comment (FriederikeHanssen) [reply lib/WorkflowMain.groovy:79]: right, I definitly think we should keep it, just make it work properly without messing with the template :)

Comment (FriederikeHanssen) [reply lib/WorkflowMain.groovy:79]: or otherwise we should raise the issue in tools and see i the template needs changing

**Pull request closed without merging**
